### PR TITLE
Use ssh-keyscan to add host to keyfile before usage

### PR DESCRIPTION
**Pull request by hyperized**
In reference to #13, a proposal to fix the issue at hand by running ssh-keyscan on the host, taking the key (hashed for security) and putting it in a temporary file, then serving it to the SSH command.

### CODE REVIEW

**Comment by hyperized** [comment]
@freekmurze Before I fix the style issues, is this the type of addition you are looking for?

**Comment by freekmurze** [comment]
Thanks! Let’s do the style fixes in a separate PR.

**Comment by hyperized** [comment]
@freekmurze I have a little problem here. I re-ran the snapshot creation for the tests, however because the temporary file name changes every run, it'll fail the tests. Does the snapshot library have a sort of wildcard option?

File: src/Ssh.php [hunk 1]
```diff
@@ -111,6 +111,10 @@ protected function getExtraOptions(): string
         if (! $this->enableStrictHostChecking) {
             $extraOptions[] = '-o StrictHostKeyChecking=no';
             $extraOptions[] = '-o UserKnownHostsFile=/dev/null';
+        } else {
+            $extraOptions[] = '-o UserKnownHostsFile='.
+                (new SshKeyScan($this->host, $this->port))
```
Comment (freekmurze) [inline src/Ssh.php]: Could you explain what `ssh-keyscan` actually does and add that to the readme?

File: src/SshKeyScan.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+class SshKeyScan
+{
+    private string $sshPort = '22';
+
+    private string $keyType = 'rsa';
+    /**
+     * @var string
+     */
+    private $path;
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Use PHP 7.4 style typing, do not use docblocks. If you need an instance variable that can't be set in the constructor or initialised properly there, make it optional.

File: src/SshKeyScan.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * @var resource
+     */
+    private $tmpFile;
+    /**
+     * @var string
+     */
+    private $tmpFilePath;
+
+    /**
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Remove this docblock

File: src/SshKeyScan.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     * SshKeyScan constructor.
+     * @param string $hostAddress
+     * @param string $sshPort
+     * @param string $keyType
+     * @throws InvalidArgumentException
+     */
+    public function __construct(string $hostAddress, ?string $sshPort = '22', ?string $keyType = 'rsa')
+    {
+        if (! $this->systemHasCapability()) {
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Rename the function to `ensureCanRunSshKeyscan`

File: src/SshKeyScan.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            throw new InvalidArgumentException('Could not execute ssh-keyscan on host filesystem');
+        }
+
+        $port = [];
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Move the building up of the options to a separate function, much like what you see in the `Ssh` class.

File: src/SshKeyScan.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        $type = [];
+        if (null !== $keyType) {
+            $type = ['-t', $keyType];
+        }
+
+        $keyscan = new Process([$this->path, ...$port, '-H', ...$type, $hostAddress]);
+        $keyscan->mustRun();
+
+        if ($keyscan->getExitCode() === 0) {
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Reverse the condition so the exception comes first. We could lose the `else` then.

File: src/SshKeyScan.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            $this->result = trim($keyscan->getOutput());
+            $this->writeContentsToTemporaryFile();
+        } else {
+            throw new InvalidArgumentException('Could not obtain host keys with ssh-keyscan');
+        }
+    }
+
+    private function writeContentsToTemporaryFile(): void
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Make this `protected`

File: src/SshKeyScan.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        $this->tmpFile = tmpfile();
+        $this->tmpFilePath = stream_get_meta_data($this->tmpFile)['uri'];
+        file_put_contents($this->tmpFilePath, $this->result);
+    }
+
+    /**
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Remove this and all similar docblocks.

File: src/SshKeyScan.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+     * @return bool
+     */
+    public function systemHasCapability(): bool
+    {
+        $which = new Process(['which', 'ssh-keyscan']);
+
+        try {
+            $which->mustRun();
```
Comment (freekmurze) [inline src/SshKeyScan.php]: Can't we use the `isSuccesfull` method instead of the `mustRun`. Would be great because it would work in the same way we use `Process` at our place in our codebases.

**Comment by hyperized** [comment]
@freekmurze I think it's all done and working (fixed the random name tests too). Let me know if you require any further changes :)

File: README.md [hunk 10]
```diff
@@ -92,10 +92,21 @@ Ssh::create('user', 'host')->usePrivateKey('/home/user/.ssh/id_rsa');
...
+By default, strict host key checking is enabled.
+
+The host key is obtained, prior to the SSH connection, by finding and running the `ssh-keyscan` utility.
+
+For more information about `ssh-keyscan`, use the `man` pages on your host system or see:
+* https://man.openbsd.org/ssh-keyscan (BSD verions)
+* https://linux.die.net/man/1/ssh-keyscan (GNU Linux versions)
+
+Warning: you will not be prompted to accept the key of the host system, it will automatically be accepted.
+You will thus lose the design feature of SSH where a new key indicates the host system has changed.
```
Comment (freekmurze) [inline README.md]: If the host key changes, will you get an error?

If not, aren't we than still susceptible to man in de middle attacks?

**Comment by freekmurze** [comment]
Thanks for this!

**Comment by freekmurze** [comment]
@hyperized The code in `SshKeyScan` was quite hard to follow, so I've almost entirely rewritten it.

> Warning: you will not be prompted to accept the key of the host system, it will automatically be accepted. You will thus lose the design feature of SSH where a new key indicates the host system has changed.

I don't get this explanation. I though that with ssh-keyscan we would add the hostkey the first time ssh connects to a certain server. I would have though that if that hostly would change, it will still give a warning.

If, with all this in place, there wouldn't be a warning when adding all this code, what's the point? It would behave exactly the same as disabling host key checking.

**Comment by hyperized** [comment]
@freekmurze 

So there's basically two events to consider here:
* The first connection to a ssh server
* Any connection to a ssh server after that

The first time if you connect to a server that's yet unknown you'll be prompted to accept the key (unless you run `ssh-keyscan` first. That issue is solved for the user with this PR.

The second or consecutive time you connect one of two scenarios may occur:
* The server key has not changed, and you will connect happily.
* The server key has changed and you will be thrown a warning.

Now I assumed (maybe that was wrong of me) that users of this package wouldn't want to be bothered because this is part of an automation system. So instead of storing the key in the users `known_hosts` file, I simply use a temporary file every time containing the latest key. This is trading off the warning of a system that may have changed with the confidence that the connection will still securely be established without a prompt.

I hope this makes sense. Perhaps we should give the users a choice about picking which of the two behaviors in the second scenario they would like.

**Comment by freekmurze** [comment]
Could you send another PR to the `v2` branch where the first time a connection is made to an ssh server the host key is put in the `known_hosts` file. If, on a subsequent connection, the hostkey has changed, it should fail. This should be the default behaviour.

If `disableStrictHostKeyChecking` was called. No checks or warnings may be given, then the user should just be connected, no questions asked.
